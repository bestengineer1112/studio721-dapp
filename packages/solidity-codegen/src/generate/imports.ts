import { AST } from 'solidity-language';

export function generateImports({
  address,
  strings,
  enumerable,
  royalties,
  delegation,
  hasAccessToken,
}: {
  address: boolean;
  strings: boolean;
  enumerable: boolean;
  royalties: boolean;
  delegation: boolean;
  hasAccessToken: boolean;
}): AST.Import[] {
  const paths: (string | AST.Import)[] = [
    ...(delegation
      ? [
          {
            path: 'gwei-slim-nft-contracts/contracts/base/ERC721Base.sol',
            names: ['ConfigSettings'],
          },
          {
            path: 'gwei-slim-nft-contracts/contracts/base/ERC721Delegated.sol',
            names: ['ERC721Delegated'],
          },
          ...(hasAccessToken
            ? ['@openzeppelin/contracts/token/ERC721/IERC721.sol']
            : []),
        ]
      : [
          '@openzeppelin/contracts/token/ERC721/ERC721.sol',
          '@openzeppelin/contracts/access/Ownable.sol',
        ]),
    '@openzeppelin/contracts/security/ReentrancyGuard.sol',
    ...(enumerable
      ? ['@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol']
      : []),
    ...(royalties ? ['@openzeppelin/contracts/interfaces/IERC2981.sol'] : []),
    '@openzeppelin/contracts/utils/Counters.sol',
    ...(address ? ['@openzeppelin/contracts/utils/Address.sol'] : []),
    ...(strings ? ['@openzeppelin/contracts/utils/Strings.sol'] : []),
  ];

  return paths.map((pathOrImport) => {
    if (typeof pathOrImport === 'string') {
      return { path: pathOrImport, names: [] };
    } else {
      return pathOrImport;
    }
  });
}
