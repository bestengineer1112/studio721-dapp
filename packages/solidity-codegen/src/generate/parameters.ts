import { ContractConfigState } from 'state';
import {
  AST,
  blockComment,
  functionDeclaration,
  structDeclaration,
  structMemberDeclaration,
  variableDeclaration,
} from 'solidity-language';

export function generateTokenParameters(
  config: Pick<ContractConfigState, 'tokenParameters'>,
): (AST.Declaration | AST.BlockComment)[] {
  return [
    blockComment({ value: 'TOKEN PARAMETERS', commentType: '/*' }),
    structDeclaration({
      name: 'TokenParameters',
      body: config.tokenParameters.map((parameter) =>
        structMemberDeclaration({
          name: parameter.name,
          typeAnnotation: parameter.type,
        }),
      ),
    }),
    variableDeclaration({
      name: 'tokenParametersMap',
      modifiers: ['private'],
      typeAnnotation: 'mapping(uint256 => TokenParameters)',
    }),
    functionDeclaration({
      name: 'tokenParameters',
      arguments: ['uint256 tokenId'],
      modifiers: ['external', 'view'],
      returns: {
        modifiers: ['memory'],
        typeAnnotation: 'TokenParameters',
      },
      body: [
        {
          type: 'returnStatement',
          expression: {
            type: 'identifier',
            value: 'tokenParametersMap[tokenId]',
          },
        },
      ],
    }),
  ];
}
