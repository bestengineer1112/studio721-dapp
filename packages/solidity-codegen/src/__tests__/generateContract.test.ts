import { ScopedAccessToken } from 'state';
import {
  ContractContext,
  getValidContractName,
  isValidSolidityIdentifier,
} from '..';
import { generateMinting } from '../generate/minting';
import { generateTokenParameters } from '../generate/parameters';
import {
  contractDeclaration,
  formatContract,
  functionCallExpression,
  identifierExpression,
  memberExpression,
  print,
} from 'solidity-language';

const defaultAccessToken: ScopedAccessToken = {
  mainnet: '',
  ropsten: '',
  rinkeby: '',
  goerli: '',
  polygon: '',
  mumbai: '',
};

const context: ContractContext = {
  msgSender: memberExpression({
    object: identifierExpression('msg'),
    member: identifierExpression('sender'),
  }),
  owner: functionCallExpression({
    callee: identifierExpression('owner'),
  }),
};

describe('contract name', () => {
  it('valid identifier', () => {
    expect(isValidSolidityIdentifier('a')).toEqual(true);
    expect(isValidSolidityIdentifier('a9')).toEqual(true);
    expect(isValidSolidityIdentifier('_')).toEqual(true);
    expect(isValidSolidityIdentifier('$')).toEqual(true);

    expect(isValidSolidityIdentifier('9')).toEqual(false);
    expect(isValidSolidityIdentifier('-9')).toEqual(false);
    expect(isValidSolidityIdentifier('a b')).toEqual(false);
  });

  it('valid contract name', () => {
    expect(getValidContractName('a')).toEqual('a');
    expect(getValidContractName('9a')).toEqual('a');
    expect(getValidContractName('99a')).toEqual('a');
    expect(getValidContractName('a b')).toEqual('ab');

    expect(getValidContractName('')).toEqual('Contract');
    expect(getValidContractName('9')).toEqual('Contract');
    expect(getValidContractName('99')).toEqual('Contract');
    expect(getValidContractName('---')).toEqual('Contract');
  });
});

describe('parameters', () => {
  it('generate', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateTokenParameters({
        tokenParameters: [
          { name: 'a', type: 'uint256' },
          { name: 'b', type: 'uint256' },
        ],
      }),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });
});

describe('minting', () => {
  const common = {
    tokenName: 'TestToken',
    shortName: 'TTKN',
    supply: 1000,
    usesIdParameter: false,
    enumerable: false,
    onlyOwnerCanMint: false,
    tokenParameters: [],
    payoutDestinations: [],
    amountAllowedForOwner: 0,
    allowlistDestinations: [],
    toggleAccessToken: false,
    mutableAccessToken: false,
    usesDelegatedContract: false,
  };

  it('base', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base + enumerable', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          enumerable: true,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base + onlyOwnerCanMint', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          onlyOwnerCanMint: true,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base + no supply', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          supply: null,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base + id', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          usesIdParameter: true,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base + id + custom max id', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          usesIdParameter: true,
          customMaxTokenId: 10000,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base + id + custom max id equal to supply', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          usesIdParameter: true,
          customMaxTokenId: common.supply,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base + accessToken', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          requireAccessToken: defaultAccessToken,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base + id + accessToken', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          usesIdParameter: true,
          requireAccessToken: defaultAccessToken,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base + id + accessToken + toggleAccessToken', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          usesIdParameter: true,
          toggleAccessToken: true,
          requireAccessToken: defaultAccessToken,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base + parameters', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          tokenParameters: [{ name: 'a', type: 'uint256' }],
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('base + id + parameters', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          usesIdParameter: true,
          tokenParameters: [{ name: 'a', type: 'uint256' }],
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('price', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          price: '0.01',
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('price + enumerable', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          price: '0.01',
          enumerable: true,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('multimint', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          multimint: 20,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('id + multimint', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          multimint: 20,
          usesIdParameter: true,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('id + multimint + accessToken', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          multimint: 20,
          usesIdParameter: true,
          requireAccessToken: defaultAccessToken,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('multimint + enumerable', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          multimint: 20,
          enumerable: true,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('multimint + price', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          multimint: 20,
          price: '0.02',
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('multimint + price + enumerable', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          multimint: 20,
          price: '0.02',
          enumerable: true,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('multimint + parameters', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          tokenParameters: [{ name: 'a', type: 'uint256' }],
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('multimint + id + parameters', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          usesIdParameter: true,
          tokenParameters: [{ name: 'a', type: 'uint256' }],
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('limit per wallet', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          limitPerWallet: 2,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('limit per wallet + multimint', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          limitPerWallet: 2,
          multimint: 2,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('owner can mint', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          amountAllowedForOwner: 5,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('owner can mint + limit per wallet', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          amountAllowedForOwner: 5,
          limitPerWallet: 2,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('owner can mint + multimint', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          multimint: 20,
          amountAllowedForOwner: 5,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });

  it('owner can mint + limit per wallet + multimint', () => {
    const contract = contractDeclaration({
      name: 'Test',
      body: generateMinting(
        {
          ...common,
          amountAllowedForOwner: 5,
          limitPerWallet: 2,
          multimint: 5,
        },
        context,
      ),
      extends: [],
    });

    expect(print(formatContract(contract))).toMatchSnapshot();
  });
});
