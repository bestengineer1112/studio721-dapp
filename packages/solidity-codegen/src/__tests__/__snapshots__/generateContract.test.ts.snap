// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minting base + accessToken 1`] = `
"contract Test {
  /** MINTING **/

  address public immutable accessTokenAddress;

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint() public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    ERC721 accessToken = ERC721(accessTokenAddress);

    require(accessToken.balanceOf(msg.sender) > 0, \\"Access token not owned\\");

    _mint(msg.sender, totalSupply());

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting base + enumerable 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  function mint() public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    _mint(msg.sender, totalSupply());
  }
}"
`;

exports[`minting base + id + accessToken + toggleAccessToken 1`] = `
"contract Test {
  /** MINTING **/

  address public immutable accessTokenAddress;

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint(uint256 id) public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    ERC721 accessToken = ERC721(accessTokenAddress);

    if (accessTokenIsActive) {
      require(accessToken.ownerOf(id) == msg.sender, \\"Access token not owned\\");
    }

    _mint(msg.sender, id);

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting base + id + accessToken 1`] = `
"contract Test {
  /** MINTING **/

  address public immutable accessTokenAddress;

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint(uint256 id) public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    ERC721 accessToken = ERC721(accessTokenAddress);

    require(accessToken.ownerOf(id) == msg.sender, \\"Access token not owned\\");

    _mint(msg.sender, id);

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting base + id + custom max id 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant MAX_TOKEN_ID = 10000;

  Counters.Counter private supplyCounter;

  function mint(uint256 id) public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(id < MAX_TOKEN_ID, \\"Invalid token id\\");

    _mint(msg.sender, id);

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting base + id + custom max id equal to supply 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint(uint256 id) public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(id < MAX_SUPPLY, \\"Invalid token id\\");

    _mint(msg.sender, id);

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting base + id + parameters 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint(uint256 id, TokenParameters calldata parameters)
    public
    nonReentrant
  {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    _mint(msg.sender, id);

    tokenParametersMap[id] = parameters;

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting base + id 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint(uint256 id) public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    _mint(msg.sender, id);

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting base + no supply 1`] = `
"contract Test {
  /** MINTING **/

  Counters.Counter private supplyCounter;

  function mint() public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    _mint(msg.sender, totalSupply());

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting base + onlyOwnerCanMint 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint() public nonReentrant onlyOwner {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    _mint(msg.sender, totalSupply());

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting base + parameters 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint(TokenParameters calldata parameters) public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    uint256 id = totalSupply();

    _mint(msg.sender, id);

    tokenParametersMap[id] = parameters;

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting base 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint() public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    _mint(msg.sender, totalSupply());

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting id + multimint + accessToken 1`] = `
"contract Test {
  /** MINTING **/

  address public immutable accessTokenAddress;

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant MAX_MULTIMINT = 20;

  Counters.Counter private supplyCounter;

  function mint(uint256[] calldata ids) public nonReentrant {
    uint256 count = ids.length;

    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() + count - 1 < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(count <= MAX_MULTIMINT, \\"Mint at most 20 at a time\\");

    ERC721 accessToken = ERC721(accessTokenAddress);

    for (uint256 i = 0; i < count; i++) {
      uint256 id = ids[i];

      require(accessToken.ownerOf(id) == msg.sender, \\"Access token not owned\\");

      _mint(msg.sender, id);

      supplyCounter.increment();
    }
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting id + multimint 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant MAX_MULTIMINT = 20;

  Counters.Counter private supplyCounter;

  function mint(uint256[] calldata ids) public nonReentrant {
    uint256 count = ids.length;

    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() + count - 1 < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(count <= MAX_MULTIMINT, \\"Mint at most 20 at a time\\");

    for (uint256 i = 0; i < count; i++) {
      uint256 id = ids[i];

      _mint(msg.sender, id);

      supplyCounter.increment();
    }
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting limit per wallet + multimint 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant MAX_MULTIMINT = 2;

  Counters.Counter private supplyCounter;

  function mint(uint256 count) public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    if (allowedMintCount(msg.sender) >= count) {
      updateMintCount(msg.sender, count);
    } else {
      revert(\\"Minting limit exceeded\\");
    }

    require(totalSupply() + count - 1 < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(count <= MAX_MULTIMINT, \\"Mint at most 2 at a time\\");

    for (uint256 i = 0; i < count; i++) {
      _mint(msg.sender, totalSupply());

      supplyCounter.increment();
    }
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting limit per wallet 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint() public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    if (allowedMintCount(msg.sender) >= 1) {
      updateMintCount(msg.sender, 1);
    } else {
      revert(\\"Minting limit exceeded\\");
    }

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    _mint(msg.sender, totalSupply());

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting multimint + enumerable 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant MAX_MULTIMINT = 20;

  function mint(uint256 count) public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() + count - 1 < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(count <= MAX_MULTIMINT, \\"Mint at most 20 at a time\\");

    for (uint256 i = 0; i < count; i++) {
      _mint(msg.sender, totalSupply());
    }
  }
}"
`;

exports[`minting multimint + id + parameters 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint(uint256 id, TokenParameters calldata parameters)
    public
    nonReentrant
  {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    _mint(msg.sender, id);

    tokenParametersMap[id] = parameters;

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting multimint + parameters 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint(TokenParameters calldata parameters) public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    uint256 id = totalSupply();

    _mint(msg.sender, id);

    tokenParametersMap[id] = parameters;

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting multimint + price + enumerable 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant MAX_MULTIMINT = 20;

  uint256 public constant PRICE = 20000000000000000;

  function mint(uint256 count) public payable nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() + count - 1 < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(count <= MAX_MULTIMINT, \\"Mint at most 20 at a time\\");

    require(
      msg.value >= PRICE * count, \\"Insufficient payment, 0.02 ETH per item\\"
    );

    for (uint256 i = 0; i < count; i++) {
      _mint(msg.sender, totalSupply());
    }
  }
}"
`;

exports[`minting multimint + price 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant MAX_MULTIMINT = 20;

  uint256 public constant PRICE = 20000000000000000;

  Counters.Counter private supplyCounter;

  function mint(uint256 count) public payable nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() + count - 1 < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(count <= MAX_MULTIMINT, \\"Mint at most 20 at a time\\");

    require(
      msg.value >= PRICE * count, \\"Insufficient payment, 0.02 ETH per item\\"
    );

    for (uint256 i = 0; i < count; i++) {
      _mint(msg.sender, totalSupply());

      supplyCounter.increment();
    }
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting multimint 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant MAX_MULTIMINT = 20;

  Counters.Counter private supplyCounter;

  function mint(uint256 count) public nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() + count - 1 < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(count <= MAX_MULTIMINT, \\"Mint at most 20 at a time\\");

    for (uint256 i = 0; i < count; i++) {
      _mint(msg.sender, totalSupply());

      supplyCounter.increment();
    }
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting owner can mint + limit per wallet + multimint 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant MAX_MULTIMINT = 5;

  Counters.Counter private supplyCounter;

  function mint(uint256 count) public nonReentrant {
    if (allowedMintCount(msg.sender) >= count) {
      updateMintCount(msg.sender, count);
    } else {
      revert(saleIsActive ? \\"Minting limit exceeded\\" : \\"Sale not active\\");
    }

    require(totalSupply() + count - 1 < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(count <= MAX_MULTIMINT, \\"Mint at most 5 at a time\\");

    for (uint256 i = 0; i < count; i++) {
      _mint(msg.sender, totalSupply());

      supplyCounter.increment();
    }
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting owner can mint + limit per wallet 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint() public nonReentrant {
    if (allowedMintCount(msg.sender) >= 1) {
      updateMintCount(msg.sender, 1);
    } else {
      revert(saleIsActive ? \\"Minting limit exceeded\\" : \\"Sale not active\\");
    }

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    _mint(msg.sender, totalSupply());

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting owner can mint + multimint 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant MAX_MULTIMINT = 20;

  Counters.Counter private supplyCounter;

  function mint(uint256 count) public nonReentrant {
    if (!saleIsActive) {
      if (allowedMintCount(msg.sender) >= count) {
        updateMintCount(msg.sender, count);
      } else {
        revert(\\"Sale not active\\");
      }
    }

    require(totalSupply() + count - 1 < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(count <= MAX_MULTIMINT, \\"Mint at most 20 at a time\\");

    for (uint256 i = 0; i < count; i++) {
      _mint(msg.sender, totalSupply());

      supplyCounter.increment();
    }
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting owner can mint 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  Counters.Counter private supplyCounter;

  function mint() public nonReentrant {
    if (!saleIsActive) {
      if (allowedMintCount(msg.sender) >= 1) {
        updateMintCount(msg.sender, 1);
      } else {
        revert(\\"Sale not active\\");
      }
    }

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    _mint(msg.sender, totalSupply());

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`minting price + enumerable 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant PRICE = 10000000000000000;

  function mint() public payable nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(msg.value >= PRICE, \\"Insufficient payment, 0.01 ETH per item\\");

    _mint(msg.sender, totalSupply());
  }
}"
`;

exports[`minting price 1`] = `
"contract Test {
  /** MINTING **/

  uint256 public constant MAX_SUPPLY = 1000;

  uint256 public constant PRICE = 10000000000000000;

  Counters.Counter private supplyCounter;

  function mint() public payable nonReentrant {
    require(saleIsActive, \\"Sale not active\\");

    require(totalSupply() < MAX_SUPPLY, \\"Exceeds max supply\\");

    require(msg.value >= PRICE, \\"Insufficient payment, 0.01 ETH per item\\");

    _mint(msg.sender, totalSupply());

    supplyCounter.increment();
  }

  function totalSupply() public view returns (uint256) {
    return supplyCounter.current();
  }
}"
`;

exports[`parameters generate 1`] = `
"contract Test {
  /** TOKEN PARAMETERS **/

  struct TokenParameters {
    uint256 a;
    uint256 b;
  }

  mapping(uint256 => TokenParameters) private tokenParametersMap;

  function tokenParameters(uint256 tokenId) external view
    returns (TokenParameters memory)
  {
    return tokenParametersMap[tokenId];
  }
}"
`;
